File: MTG Deck Editor/src/editor/database/symbol/VariableSymbol.java
package editor.database.symbol;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;

import editor.database.characteristics.ManaType;

/**
 * This class represents a symbol representing a variable amount of generic mana
 * using the variables X, Y or Z.
 * 
 * @author Alec Roelke
 */
public class VariableSymbol extends ManaSymbol
{
	/**
	 * Map of variable names onto their corresponding symbols.
	 */
	public static final Map<String, VariableSymbol> SYMBOLS = Collections.unmodifiableMap(
			Arrays.asList('X', 'Y', 'Z').stream().collect(Collectors.toMap(String::valueOf, VariableSymbol::new)));
	
	/**
	 * Get the VariableSymbol corresponding to the given String.
	 * 
	 * @param x the String to look up
	 * @return the VariableSymbol corresponding to the String
	 * @throws IllegalArgumentException if the String doesn't correspond to a variable symbol
	 */
	public static VariableSymbol parseVariableSymbol(String x)
	{
		VariableSymbol symbol = tryParseVariableSymbol(x);
		if (symbol == null)
			throw new IllegalArgumentException('"' + x + "\" is not a variable symbol");
		return symbol;
	}
	
	/**
	 * Get the VariableSymbol corresponding to the given String.
	 * 
	 * @param x the String to look up
	 * @return the VariableSymbol corresponding to the String, or null if there
	 * is no such symbol.
	 */
	public static VariableSymbol tryParseVariableSymbol(String x)
	{
		return SYMBOLS.get(x.toUpperCase());
	}
	
	/**
	 * Variable name for this VariableSymbol.
	 */
	private final char var;
	
	/**
	 * Create a new VariableSymbol with the corresponding variable.
	 * 
	 * @param v variable name for the new VariableSymbol
	 */
	private VariableSymbol(char v)
	{
		super(Character.toLowerCase(v) + "_mana.png", String.valueOf(Character.toUpperCase(v)), 0);
		var = Character.toUpperCase(v);
	}

	/**
	 * {@inheritDoc}
	 * {@link ManaType#COLORLESS} is mapped to 0.5 and every other color is mapped to
	 * 0.
	 */
	@Override
	public Map<ManaType, Double> colorWeights()
	{
		return createWeights(new ColorWeight(ManaType.COLORLESS, 0.5));
	}
	
	@Override
	public int compareTo(ManaSymbol other)
	{
		if (other instanceof VariableSymbol)
			return var - ((VariableSymbol)other).var;
		else
			return super.compareTo(other);
	}
}

File: src/main/scala/editor/database/symbol/VariableSymbol.scala
package editor.database.symbol

import editor.database.attributes.ManaType

import scala.jdk.CollectionConverters._
import scala.jdk.OptionConverters._

/**
 * Companion to [[VariableSymbol]] containing all the possible variables used by the game and functions for parsing
 * them from strings.
 *
 * @author Alec Roelke
 */
object VariableSymbol {
  @deprecated val SYMBOLS = VariableSymbolGenerator.values.asJava
  @deprecated def parseVariableSymbol(s: String) = VariableSymbolGenerator.values(s(0).toUpper)
  @deprecated def tryParseVariableSymbol(s: String) = VariableSymbolGenerator.parse(s).toJava
}

/**
 * A symbol that can be paid using any amount of mana, indicating using a letter like X.  Its color intensity is 0 for all mana types, as
 * any type of mana can be spent to pay for it, and its text representation is the letter used as the variable.
 * 
 * @constructor create a new variable symbol using a particular variable name
 * @param variable letter to use as the variable name
 * 
 * @author Alec Roelke
 */
class VariableSymbol private[symbol](private val variable: Char) extends ManaSymbol(s"${variable.toLower}_mana.png", variable.toString.toUpperCase, 0) {
  override def colorIntensity = ManaSymbol.createIntensity(Map(ManaType.COLORLESS -> 0.5))

  override def compareTo(other: ManaSymbol) = other match {
    case s: VariableSymbol => variable.toUpper - s.variable.toUpper
    case _ => super.compareTo(other)
  }
}